go/parquet/compress/compress.go

```diff
@@ -0,0 +1,153 @@
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package compress contains the interfaces and implementations for handling compression/decompression
+// of parquet data at the column levels.
+package compress
+
+import (
+	"compress/flate"
+	"io"
+	"io/ioutil"
+
+	"github.com/apache/arrow/go/parquet/internal/gen-go/parquet"
+)
+
+// Compression is an alias to the thrift compression codec enum type for easy use
+type Compression parquet.CompressionCodec
+
+func (c Compression) String() string {
+	return parquet.CompressionCodec(c).String()
+}
+
+// DefaultCompressionLevel will use flate.DefaultCompression since many of the compression libraries
+// use that to denote "use the default".
+const DefaultCompressionLevel = flate.DefaultCompression
+
+// Codecs is a useful struct to provide namespaced enum values to use for specifying the compression type to use
+// which make for easy internal swapping between them and the thrift enum since they are initialized to the same
+// constant values.
+var Codecs = struct {
+	Uncompressed Compression
+	Snappy       Compression
+	Gzip         Compression
+	// LZO is unsupported in this library as I haven't yet found a good implementation of it for golang.
+	Lzo    Compression
+	Brotli Compression
+	// LZ4 unsupported in this library due to problematic issues between the Hadoop LZ4 spec vs regular lz4
+	// see: http://mail-archives.apache.org/mod_mbox/arrow-dev/202007.mbox/%3CCAAri41v24xuA8MGHLDvgSnE+7AAgOhiEukemW_oPNHMvfMmrWw@mail.gmail.com%3E
+	Lz4  Compression
+	Zstd Compression
+}{
+	Uncompressed: Compression(parquet.CompressionCodec_UNCOMPRESSED),
+	Snappy:       Compression(parquet.CompressionCodec_SNAPPY),
+	Gzip:         Compression(parquet.CompressionCodec_GZIP),
+	Lzo:          Compression(parquet.CompressionCodec_LZO),
+	Brotli:       Compression(parquet.CompressionCodec_BROTLI),
+	Lz4:          Compression(parquet.CompressionCodec_LZ4),
+	Zstd:         Compression(parquet.CompressionCodec_ZSTD),
+}
+
+// Codec is an interface which is implemented for each compression type in order to make the interactions easy to
+// implement. Most consumers won't be calling GetCodec directly.
+type Codec interface {
+	// NewReader provides a reader that wraps a stream with compressed data to stream the uncompressed data
+	NewReader(io.Reader) io.ReadCloser
+	// NewWriter provides a wrapper around a write stream to compress data before writing it.
+	NewWriter(io.Writer) io.WriteCloser
+	// NewWriterLevel is like NewWrapper but allows specifying the compression level
+	NewWriterLevel(io.Writer, int) (io.WriteCloser, error)
```

Comment:
API comment.  It is worth exposing compression level in the API, an laternative would have be a member used at construction time?

Reply:
So the way I built this was to minimize object creation, so the map contains objects which are stateless to ensure we have no race conditions, the objects themselves know how to construct the appropriate encoders and decoders, that's why I exposed it like this since not all of the libraries allow configuring the compression level *after* constructing the encoder. Most of them take the compression level as an argument to their `NewWriter` functions etc. which is what I modeled this after.